zenith_utils/src/postgres_backend.rs

```diff
@@ -0,0 +1,147 @@
+// Server-side Postgres connection, as limited as we need.
+
+use crate::pq_proto::{BeMessage, FeMessage, FeStartupMessage, StartupRequestCode};
+use anyhow::bail;
+use anyhow::Result;
+use bytes::Bytes;
+use log::*;
+use std::io;
+use std::io::{BufReader, BufWriter, Write};
+use std::net::TcpStream;
+
+pub trait Handler {
+    // handle single query
+    fn process_query(&mut self, pgb: &mut PostgresBackend, query_string: Bytes) -> Result<()>;
+    // called on startup packet receival, allows to process params
+    fn startup(&mut self, _pgb: &mut PostgresBackend, _sm: &FeStartupMessage) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub struct PostgresBackend {
+    socket: TcpStream,
+    // replication.rs wants to handle reading on its own in separate thread, so
+    // wrap in Option to be able to take and transfer the BufReader. Ugly, but I
+    // have no better ideas.
+    stream_in: Option<BufReader<TcpStream>>,
+    stream: BufWriter<TcpStream>,
+    init_done: bool,
+}
+
+impl PostgresBackend {
+    pub fn new(socket: TcpStream) -> Self {
+        PostgresBackend {
+            stream_in: Some(BufReader::new(socket.try_clone().unwrap())),
+            stream: BufWriter::new(socket.try_clone().unwrap()),
```

Comment:
Should we handle errors from `try_clone`?
I assumed it would be just copying the integer file descriptor, but it seems like [it makes a syscall](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/unix/fd.rs#L267)

Reply:
This is one of those rare times where I think panicking is probably the right thing. If `dup` fails, it probably means we're running out of file descriptors, and there's really nothing we can do short of killing the daemon and letting it restart.

Reply:
> there's really nothing we can do short of killing the daemon

this way we will impact existing connections instead of throwing connection error only for new ones. +1 on handling it gracefully.

Reply:
It's possible to bump RLIMIT_NOFILE of a running process using [prlimit](https://man7.org/linux/man-pages/man2/setrlimit.2.html).

There's also a tool which shares the syscall's name.

Reply:
Done graceful handling.